packages/playwright-test/src/reporters/azure.ts

```diff
@@ -0,0 +1,481 @@
+
+    this.createRun(this.testRunTitle).then(run => {
+      if (run) {
+        this.runId = run.id;
+        this.log(colors.green(`Using run ${this.runId} to publish test results`));
+      } else {
+        this.isDisabled = true;
+      }
+    });
+  }
+
+  async onTestEnd(test: TestCase, testResult: TestResult): Promise<void> {
+    await this.awaitForRunId().catch(e => {
+      this.log(colors.red(e));
+    });
+    if (this.isDisabled)
+      return;
+
+    this.logTestItem(test, testResult);
+    await this.publishCaseResult(test, testResult);
+  }
+
+  async onEnd(): Promise<void> {
+    await this.awaitForRunId().catch(e => {
+      this.log(colors.red(e));
+    });
+    if (this.isDisabled)
+      return;
+
+    let prevCount = this.resultsToBePublished.length;
+    while (this.resultsToBePublished.length > 0) {
+      // need wait all results to be published
+      if (prevCount > this.resultsToBePublished.length) {
+        this.log(
+            colors.gray(`Waiting for all results to be published. Remaining ${this.resultsToBePublished.length} results`)
+        );
+        prevCount--;
+      }
+      await new Promise(resolve => setTimeout(resolve, 250));
+    }
+
+    if (this.publishedResultsCount === 0 && !this.runId) {
+      this.log('No testcases were matched. Ensure that your tests are declared correctly.');
+      return;
+    }
+
+    try {
+      const runUpdate: TestInterfaces.RunUpdateModel = {
+        state: 'Completed'
+      };
+      if (!this.testApi)
+        this.testApi = await this.connection.getTestApi();
+      const runUpdatedResponse = await this.testApi.updateTestRun(runUpdate, this.projectName, this.runId as number);
+      this.log(colors.green(`Run ${this.runId} - ${runUpdatedResponse.state}`));
+    } catch (err) {
+      this.log(`Error on completing run ${err as string}`);
+    }
+  }
+
+  private log(message?: any) {
+    if (this.logging)
+      console.log(colors.magenta(`azure: ${message}`));
+
+  }
+
+  private getCaseIds(test: TestCase): string {
+    const regexp = /\[([\d,]+)\]/;
+    const results = regexp.exec(test.title);
+    if (results && results.length === 2)
+      return results[1];
+
+    return '';
+  }
+
+  private logTestItem(test: TestCase, testResult: TestResult) {
+    const map = {
+      failed: colors.red(`Test ${test.title} - ${testResult.status}`),
+      passed: colors.green(`Test ${test.title} - ${testResult.status}`),
+      skipped: colors.blue(`Test ${test.title} - ${testResult.status}`),
+      pending: colors.blue(`Test ${test.title} - ${testResult.status}`),
+      disabled: colors.gray(`Test ${test.title} - ${testResult.status}`),
+      timedOut: colors.gray(`Test ${test.title} - ${testResult.status}`)
+    };
+    if (testResult.status)
+      this.log(map[testResult.status]);
+
+  }
+
+  private async createRun(runName: string): Promise<TestInterfaces.TestRun> {
+    return new Promise<TestInterfaces.TestRun>(async (resolve, reject) => {
```

Comment:
just thinking if a rewrite here makes sense without using "new Promise". I think it would make it easier to follow.